diff -urN org/sudoers new/sudoers
--- org/sudoers	2018-02-15 20:30:54.503155177 +0000
+++ new/sudoers	2018-02-15 20:32:34.911159166 +0000
@@ -76,7 +76,7 @@
 #
 # Defaults   env_keep += "HOME"
 
-Defaults    secure_path = /sbin:/bin:/usr/sbin:/usr/bin
+Defaults    secure_path = /sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin
 
 ## Next comes the main part: which users can run what software on 
 ## which machines (the sudoers file can be shared between multiple
